### PR TITLE
Javadoc examples for azconfig

### PR DESCRIPTION
**Pull request by alzimmermsft**
Fixes issue #3322 

Added JavaDocs to ConfigurationAsyncClient and ConfigurationAsyncClientBuilder

### CODE REVIEW

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClientBuilder.java [hunk 1]
```diff
@@ -26,7 +26,50 @@
...
+ *     construct a HttpPipeline that will be used by the client to interact with the service; a new pipeline will be
+ *     constructed every build.
+ *
+ * <pre>
+ * {@code ConfigurationClientCredentials credentials = new ConfigurationClientCredentials(connectionString);
+ *    ConfigurationAsyncClient.builder()
+ *        .credentials(credentials)
+ *        .build();}
+ * </pre>
+ * </p>
```
Comment (conniey) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClientBuilder.java]: This closing tag `</p>` is not necessary because the paragraph is followed by a `<pre>` and shows as an error when generating Javadocs.
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/p

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClientBuilder.java [hunk 2]
```diff
@@ -26,7 +26,50 @@
...
+ *
+ * <pre>
+ * {@code String serviceEndpoint = <App-Configuration-URL>;
+ *    HttpPipeline pipeline = new HttpPipeline(<policies>);
+ *    ConfigurationAsyncClient.builder()
+ *        .pipeline(pipeline)
+ *        .serviceEndpoint(serviceEndpoint)
+ *        .build();}
+ * </pre>
+ * </p>
```
Comment (conniey) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClientBuilder.java]: Not necessary. See comment above

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClient.java [hunk 3]
```diff
@@ -27,6 +27,73 @@
...
+ *
+ * Update an existing Configuration Setting
+ * <pre>
+ *     {@code
+ *     client.updateSetting("prodDBConnection", "<updated_db_connection").subscribe(result -> {
+ *         ConfigurationSetting setting = result.value();
+ *         System.out.println(String.format("Key: %s, Value: %s", setting.key(), setting.value());
+ *     });
+ *
+ *     client.setSetting("testDBConnection", "<updated_db_connection>").subscribe(result -> {
```
Comment (conniey) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClient.java]: Might be worthwhile to describe that `setSetting` can add or update a configuration setting. From the heading "Update an existing Configuration Setting", it seems like you need to perform both an `updateSetting` followed by a `setSetting`.

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClientBuilder.java (hunk 1 above)
Comment (JonathanGiles) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClientBuilder.java]: I would prefer to see the `</p>` placed before the opening `<pre>` code block. This should work as expected.

Comment (JonathanGiles) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClientBuilder.java]: Also, be sure to generate the javadoc on your machine to ensure everything is aligned appropriately. I typically don't put multi-line code starting on the same line as `{@code` - I normally start on a new line after that.

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClient.java [hunk 4]
```diff
@@ -27,6 +27,73 @@
  * Client that contains all the operations for {@link ConfigurationSetting ConfigurationSettings} in Azure Configuration
  * Store.
  *
+ * <p>
+ *     Azure Configuration Store operations allowed by the client are adding, retrieving, updating, and deleting
+ *     ConfigurationSettings, additionally more complex operations of listing settings or revision of a setting based
+ *     on a filter are supported as well.
+ * </p>
```
Comment (JonathanGiles) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClient.java]: Just a general formatting preference - Put 'Azure' directly after the opening `<p>` element, and similarly, put the closing element at the end of the sentence rather than on a new line. Then, don't do any indentation like you have here - just a single space between the asterisk and the text.

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClient.java [hunk 5]
```diff
@@ -27,6 +27,73 @@
  * Client that contains all the operations for {@link ConfigurationSetting ConfigurationSettings} in Azure Configuration
  * Store.
  *
+ * <p>
+ *     Azure Configuration Store operations allowed by the client are adding, retrieving, updating, and deleting
+ *     ConfigurationSettings, additionally more complex operations of listing settings or revision of a setting based
```
Comment (conniey) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClient.java]: Adding a `{@link ConfigurationSetting ConfigurationSettings}` would make it easier to click on things and figure out what they are as I'm reading the docs.

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClient.java [hunk 6]
```diff
@@ -27,6 +27,73 @@
...
+ * Construct the client
+ * <pre>
+ *     {@code
+ *     String connectionString = <connection_string>;
+ *     ConfigurationClientCredentials credentials = new ConfigurationClientCredentials(connectionString);
+ *     ConfigurationAsyncClient client = ConfigurationAsyncClient.builder()
+ *          .credentials(credentials)
+ *          .build();
+ *     }
+ * </pre>
```
Comment (JonathanGiles) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClient.java]: Putting `{@code` inside a pre block is the wrong way around, I believe. Generate the JavaDoc and check. Also, don't indent the code like this, start with no indentation and then indent as relevant.

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClient.java (hunk 5 above)
Comment (alzimmermsft) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClient.java]: Do we want to @link all class and method references or are you asking for this particular spot?

Comment (alzimmermsft) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClient.java]: Do we want to `@link` all class and method references or do you want this particular instance linked?

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClient.java (hunk 6 above)
Comment (alzimmermsft) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClient.java]: Reading up on it there is a purpose and valid use case for `@code` in a pre block but this wasn't the right one.

**Comment by JonathanGiles** [comment]
When you've made this change, let's sync up and do a review of the generated JavaDoc output. Schedule 30 minutes of my time for whenever suits you.

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClient.java [hunk 7]
```diff
@@ -25,7 +25,16 @@
 
 /**
  * Client that contains all the operations for {@link ConfigurationSetting ConfigurationSettings} in Azure Configuration
- * Store.
+ * Store. Operations allowed by the client are adding, retrieving, updating, and deleting ConfigurationSettings, and
+ * listing settings or revision of a setting based on a filter.
+ *
+ * <p><strong>Construct the client, for additional ways to build a client see {@link ConfigurationAsyncClientBuilder here}.</strong></p>
```
Comment (JonathanGiles) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClient.java]: I would changing this line. You should have only 'Instantiating an Asynchronous Configuration Client' or similar in bold. The rest of the text above should go in a new paragraph beneath the code block

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClient.java [hunk 8]
```diff
@@ -78,6 +94,17 @@ public static ConfigurationAsyncClientBuilder builder() {
      * Adds a configuration value in the service if that key and label does not exist. The label value of the
      * ConfigurationSetting is optional.
      *
+     * <p><strong>Code Samples</strong></p>
+     * <pre>
+     * ConfigurationSetting setting = new ConfigurationSetting().key("prodDBConnection")
+     *     .label("westUS")
+     *     .value("db_connection");
+     *
+     * client.addSetting(setting).subscribe(response -&gt; {
```
Comment (JonathanGiles) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClient.java]: I would prefer to see this kind of code sample with the ConfigurationSetting passed directly into addSetting, without the intermediate `setting` variable. Ping me to get a good style.

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClient.java [hunk 9]
```diff
@@ -78,6 +94,17 @@ public static ConfigurationAsyncClientBuilder builder() {
      * Adds a configuration value in the service if that key and label does not exist. The label value of the
      * ConfigurationSetting is optional.
      *
+     * <p><strong>Code Samples</strong></p>
+     * <pre>
```
Comment (JonathanGiles) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClient.java]: Newline before the `pre` - try to enforce consistency on yourself in every aspect! :-)

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClient.java [hunk 10]
```diff
@@ -98,6 +125,19 @@ public static ConfigurationAsyncClientBuilder builder() {
...
+     * client.setSetting("prodDBConnection", "db_connection").subscribe(response -&gt; {
+     *     ConfigurationSetting result = response.value();
+     *     System.out.printf("Key: %s, Value: %s", result.key(), result.value());
+     * });
+     *
+     * client.setSetting("prodDBConnection", "updated_db_connection").subscribe(response -&gt; {
+     *     ConfigurationSetting result = response.value();
+     *     System.out.printf("Key: %s, Value: %s", result.key(), result.value());
+     * });</pre>
```
Comment (JonathanGiles) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClient.java]: In scenarios where you are demonstrating more than just how to use the code, such as here, you should add additional text that states the value of the `prodDBConnection` setting before either call, in the middle, and after the second call.

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClient.java [hunk 11]
```diff
@@ -117,6 +157,24 @@ public static ConfigurationAsyncClientBuilder builder() {
...
+     * ConfigurationSetting setting = new ConfigurationSetting().key("prodDBConnection")
+     *     .label("westUS")
+     *     .value("db_connection");
+     *
+     * client.setSetting(setting).subscribe(response -&gt; {
+     *     ConfigurationSetting result = response.value();
+     *     System.out.printf("Key: %s, Value: %s", result.key(), result.value());
+     * });
+     *
+     * setting = setting.value("updated_db_connection");
```
Comment (JonathanGiles) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClient.java]: This is not good code - ConfigurationSetting is a mutable class, so you do not need to reassign it to `setting` again

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClient.java [hunk 12]
```diff
@@ -251,6 +368,13 @@ public static ConfigurationAsyncClientBuilder builder() {
      * Fetches the configuration settings that match the {@code options}. If {@code options} is {@code null}, then all
      * the {@link ConfigurationSetting configuration settings} are fetched with their current values.
      *
+     * <p><strong>Code Samples</strong></p>
+     *
+     * <pre>
+     * client.listSettings(new SettingSelector().key("prodDBConnection")).subscribe(setting -&gt; {
```
Comment (JonathanGiles) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClient.java]: When you move the setting selector inline like this, in my opinion you should move the `.subscribe` to the next line.

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClientBuilder.java [hunk 13]
```diff
@@ -26,7 +26,30 @@
 import java.util.Objects;
 
 /**
- * Provides configuration options for instances of {@link ConfigurationAsyncClient}.
+ * Builds instances of {@link ConfigurationAsyncClient} based on the configuration options stored in the builder when
+ * {@link ConfigurationAsyncClientBuilder#build() build} is called. Preparing the builder has two primary options, using
+ * {@link ConfigurationClientCredentials} or using a {@link HttpPipeline} and the service endpoint.
+ *
+ * <p>ConfigurationClientCredentials has the service endpoint and authorization information built into it. Use
+ * {@link ConfigurationAsyncClientBuilder#credentials(ConfigurationClientCredentials) credentials} to configure the builder.</p>
```
Comment (JonathanGiles) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClientBuilder.java]: Let's sync up about these first two paragraphs. I think there needs to be a few tweaks. Notably:

1. 'Builds' is not a good start to the paragraph. We should say something like "This class provides a fluent builder API to help aid the configuration and instantiation of the ...."
2. I don't like the "Preparing the builder has two primary options" sentence. It is hard to parse, and too detailed for a first paragraph.
3. I think if you put yourself in the shoes of a newcomer to this API, the second paragraph above would provide far too little actionable information.

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClientBuilder.java [hunk 14]
```diff
@@ -26,7 +26,29 @@
...
+ * </pre>
+ *
+ * <p>{@link HttpPipeline} gives the builder the access credentials it requires, set the HttpPipeline with {@link ConfigurationAsyncClientBuilder#pipeline(HttpPipeline) pipeline}.
+ * Using a HttpPipeline requires additional setup but allows for finer control on how the ConfigurationAsyncClient it built.
+ * Additionally, the service endpoint is also required, set the service endpoint with {@link ConfigurationAsyncClientBuilder#serviceEndpoint(String) serviceEndpoint}.</p>
+ *
+ * <pre>
+ * ConfigurationAsyncClient.builder()
+ *     .pipeline(new HttpPipeline(policies))
+ *     .serviceEndpoint(serviceEndpoint)
```
Comment (JonathanGiles) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClientBuilder.java]: Is this the consistently applied name (`serviceEndpoint`)? Check with Key Vault.

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClientBuilder.java [hunk 15]
```diff
@@ -13,23 +13,22 @@
...
  *     .build();
  * </pre>
  *
- * <p>HttpPipeline is constructed using the authorization information but doesn't have the service endpoint built into
- * it. Use {@link ConfigurationClientBuilder#pipeline(HttpPipeline) pipeline} and {@link ConfigurationClientBuilder#serviceEndpoint(String) serviceEndpoint}
- * to configure the builder. The HttpPipeline requires additional configuration over using ConfigurationClientCredentials
- * but it allows for finer control on how the ConfigurationClient is built.</p>
+ * <p>{@link HttpPipeline} gives the builder the access credentials it requires, set the HttpPipeline with {@link ConfigurationClientBuilder#pipeline(HttpPipeline) pipeline}.
+ * Using a HttpPipeline requires additional setup but allows for finer control on how the ConfigurationClient it built.
+ * Additionally, the service endpoint is also required, set the service endpoint with {@link ConfigurationClientBuilder#serviceEndpoint(String) serviceEndpoint}.</p>
```
Comment (JonathanGiles) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClientBuilder.java]: This paragraph is hard to read.

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClientBuilder.java (hunk 14 above)
Comment (alzimmermsft) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClientBuilder.java]: KeyVault's equivalent is url, which I think it the better term to use as we reference URL when talking about it and the backing type is URL.

Comment (conniey) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClientBuilder.java]: I think "serviceEndpoint" makes it easier to interpret rather than just "url" (what URI are we referring to?) When you copy these values from portal.azure.com services like App Config, Storage, Cognitive Services call this value: "Endpoint" and "Service Endpoint"

Comment (JonathanGiles) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClientBuilder.java]: I would like to standardise here, yes. My initial instinct is to go with `endpoint`. I'm not sure that `serviceEndpoint` adds a lot more contextual information (that is, what other endpoint would we be referring to?). Can someone also please confer with the other languages to see where they have landed in this naming. Any other thoughts? // cc @g2vinay

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClientBuilder.java (hunk 15 above)
Comment (alzimmermsft) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClientBuilder.java]: Reworded it.

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClientBuilder.java (hunk 14 above)
Comment (alzimmermsft) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClientBuilder.java]: Appears .NET and Python use connectionString/connection_string and they extract the URL from that.

Comment (conniey) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClientBuilder.java]: Hm. That's one way to authenticate with Key Vault. You can also use a service principal, in which case there is no such thing as a connection string. You would use the keyvault URL, client ID and client secret.

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClient.java [hunk 16]
```diff
@@ -78,6 +100,18 @@ public static ConfigurationAsyncClientBuilder builder() {
...
+     *
+     * <p>Add a setting with the key "prodDBConnection", label "westUS", and value "db_connection".</p>
+     *
+     * <pre>
+     * client
+     *     .addSetting(new ConfigurationSetting().key("prodDBConnection").label("westUS").value("db_connection"))
+     *     .subscribe(response -&gt; {
+     *         ConfigurationSetting result = response.value();
+     *         System.out.printf("Key: %s, Value: %s", result.key(), result.value());
+     *     });</pre>
```
Comment (JonathanGiles) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClient.java]: Different code styles between here and line 81 - here you put 'addSetting' on a new line, whereas on line 81 you don't.

Comment (JonathanGiles) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClient.java]: In my opinion, always have the first operation on the same line, so the code blocks in your JavaDoc that look like this should be improved.

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClient.java [hunk 17]
```diff
@@ -161,6 +248,18 @@ public static ConfigurationAsyncClientBuilder builder() {
      *
      * If {@link ConfigurationSetting#etag() etag} is specified, the configuration value is only updated if it matches.
      *
+     * <p><strong>Code Samples</strong></p>
+     *
+     * <p>Update the setting with the key-label pair "prodDBConnection"-"westUS" to have the value "updated_db_connection".</p>
+     *
+     * <pre>
+     * client
+     *     .updateSetting(new ConfigurationSetting().key("prodDBConnection").label("westUS").value("db_connection"))
```
Comment (JonathanGiles) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClient.java]: Your text on line 253 says the value will be updated to "updated_db_connection", but that isn't what you're passing in to the method call.

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClient.java [hunk 18]
```diff
@@ -196,6 +306,18 @@ public static ConfigurationAsyncClientBuilder builder() {
     /**
      * Attempts to get the ConfigurationSetting given the {@code key}, optional {@code label}.
      *
+     * <p><strong>Code Samples</strong></p>
+     *
+     * <p>Retrieve the setting with the key-label "prodDBConnection"-"westUS".</p>
+     *
+     * <pre>
+     * client
+     *     .getSetting(new ConfigurationSetting().key("prodDBConnection").label("westUS").value("db_connection"))
```
Comment (JonathanGiles) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClient.java]: You are setting a `value` in this request erroneously?

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java [hunk 19]
```diff
@@ -161,6 +171,8 @@ public static ConfigurationClientBuilder builder() {
      *
      * <p><strong>Code Samples</strong></p>
      *
+     * <p>Update a setting with the key "prodDBConnection" to have the value "updated_db_connection".</p>
```
Comment (JonathanGiles) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java]: This doesn't match what is shown in the code below. Firstly, there is a typo in the code `prodDCConnection`, and secondly, it says `db_connection` rather than `updated_db_connection`

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java [hunk 20]
```diff
@@ -185,12 +197,11 @@ public static ConfigurationClientBuilder builder() {
...
-     * <pre>
-     * ConfigurationSetting setting = new ConfigurationSetting().key("prodDBConnection")
-     *     .label("westUS")
-     *     .value("db_connection");
+     * <p>Update the setting with the key-label pair "prodDBConnection"-"westUS" to have the value "updated_db_connection".</p>
      *
-     * ConfigurationSetting result = client.updateSetting(setting);
+     * <pre>
+     * ConfigurationSetting result = client
+     *     .updateSetting(new ConfigurationSetting().key("prodDBConnection").label("westUS").value("db_connection"));
```
Comment (JonathanGiles) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java]: Same issue here with value not matching the text in the description. Feels like copy/paste fails :)

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClient.java (hunk 16 above)
Comment (alzimmermsft) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClient.java]: All code snippets changed

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java (hunk 20 above)
Comment (alzimmermsft) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java]: Yes, yes there were a lot of those...
